### PR TITLE
[clang-tidy] Option to ignore anonymous namespaces in avoid-non-const-global-variables

### PR DESCRIPTION
**Pull request by pascalj**
Add an option to ignore warnings for cppcoreguidelines avoid-non-const-global-variables.

Understandably, the core guidelines discourage non const global variables, even at the TU level (see https://github.com/isocpp/CppCoreGuidelines/issues/2195). However, having a small TU with an interface that uses a non const variable from an anonymous namespace can be a valid choice.

This adds an option that disables the warning just for anonymous namespaces, i.e. at the file level. The default is still to show a warning, just as before.

### CODE REVIEW

File: clang-tools-extra/docs/clang-tidy/checks/cppcoreguidelines/avoid-non-const-global-variables.rst [hunk 1]
```diff
@@ -41,3 +41,25 @@ The variables ``a``, ``c``, ``c_ptr1``, ``c_const_ptr`` and ``c_reference``
 will all generate warnings since they are either a non-const globally accessible
 variable, a pointer or a reference providing global access to non-const data
 or both.
+
+Options
+-------
+
+.. option:: AllowAnonymousNamespace
+
+   When set to `true` (default is `false`), non-const variables in anonymous namespaces will not generate a warning.
```
Comment (EugeneZelenko) [inline clang-tools-extra/docs/clang-tidy/checks/cppcoreguidelines/avoid-non-const-global-variables.rst:50]: Default value is usually stated in dedicated sentence at the end of paragraph.

**Review by PiotrZSL — CHANGES_REQUESTED** [review]
- storeOptions is missing
- release notes need to be updated about added new option
- tests need some love (reduce amount of changes)
- what about "static" non const global variables 

I'm fine for having an options that control behavior of the check.
But please note that there are other aspects of "I.2: Avoid non-const global variables" rule that anonymous namespace or static does not solve. Like race-conditions in multithread env. Or even to force developer not to use global objects to pass data between functions.

Even that I see many times that warnings from this check are being nolinted, I still think that it's doing good job.

**Comment by carlosgalvezp** [comment]
I have a bit the feeling that users should use NOLINT* here instead of having a global option that is not seen in the code. Anonymous namespaces do not solve many of the remaining issues. A NOLINT in the code can grab the attention of a future developer and prompt a refactoring to avoid the pattern.

I don't have a strong opinion so if this is really wanted I won't object.

File: clang-tools-extra/test/clang-tidy/checkers/cppcoreguidelines/avoid-non-const-global-variables.cpp [hunk 2]
```diff
@@ -1,29 +1,39 @@
-// RUN: %check_clang_tidy %s cppcoreguidelines-avoid-non-const-global-variables %t
+// RUN: %check_clang_tidy %s -check-suffix=DEFAULT cppcoreguidelines-avoid-non-const-global-variables %t
+// RUN: %check_clang_tidy %s -check-suffix=NAMESPACE cppcoreguidelines-avoid-non-const-global-variables %t -- \
```
[the remark this answers is not in the document]
Comment (pascalj) [reply clang-tools-extra/test/clang-tidy/checkers/cppcoreguidelines/avoid-non-const-global-variables.cpp]: Thanks for this hint, it made the checks much easier. It now tests only for the cases where the configuration differs.

**Comment by pascalj** [comment]
Thanks for your feedback!

>     * what about "static" non const global variables

Good point, forgot about these. If both are allowed, it is more about the internal linkage than it is about the namespace. I renamed the option to `AllowInternalLinkage` and permit variables in anonymous namespaces and static global variables, since they're equivalent (AFAICT). 

> I'm fine for having an options that control behavior of the check. But please note that there are other aspects of "I.2: Avoid non-const global variables" rule that anonymous namespace or static does not solve. Like race-conditions in multithread env. Or even to force developer not to use global objects to pass data between functions.
> 
> Even that I see many times that warnings from this check are being nolinted, I still think that it's doing good job.

I completely agree with both you and @carlosgalvezp: this is neither an option that should be enabled by default, nor should people prefer it over nolinting without some thought. However, I see not much harm in giving someone the choice to do so. In the end, all of clang-tidy's checks are optional to some degree.

Having said that: if the maintainers prefer not to have this option, this is also fine with me.

**Review by PiotrZSL — APPROVED** [review]
LGTM

File: clang-tools-extra/clang-tidy/cppcoreguidelines/AvoidNonConstGlobalVariablesCheck.cpp [hunk 3]
```diff
@@ -7,21 +7,30 @@
...
 #include "clang/ASTMatchers/ASTMatchers.h"
 
 using namespace clang::ast_matchers;
 
 namespace clang::tidy::cppcoreguidelines {
 
+AvoidNonConstGlobalVariablesCheck::AvoidNonConstGlobalVariablesCheck(
+    StringRef Name, ClangTidyContext *Context)
+    : ClangTidyCheck(Name, Context),
+      AllowInternalLinkage(Options.get("AllowInternalLinkage", false)) {}
```
Comment (SimplyDanny) [inline clang-tools-extra/clang-tidy/cppcoreguidelines/AvoidNonConstGlobalVariablesCheck.cpp:20]: Option name could be a constant to avoid duplication.

**Review by SimplyDanny — APPROVED** [review]

File: clang-tools-extra/test/clang-tidy/checkers/cppcoreguidelines/avoid-non-const-global-variables.cpp
Comment (5chmidti) [inline clang-tools-extra/test/clang-tidy/checkers/cppcoreguidelines/avoid-non-const-global-variables.cpp:1]: FYI: You could remove the explicit check-nots here:

```c++
// RUN: %check_clang_tidy %s -check-suffixes=,INTERNAL-LINKAGE cppcoreguidelines-avoid-non-const-global-variables %t
// RUN: %check_clang_tidy %s cppcoreguidelines-avoid-non-const-global-variables %t -- \
// RUN: -config="{CheckOptions: {cppcoreguidelines-avoid-non-const-global-variables.AllowInternalLinkage : 'true'}}"


namespace {
int nonConstAnonymousNamespaceInt = 0;
// CHECK-MESSAGES-INTERNAL-LINKAGE: :[[@LINE-1]]:5: warning: variable 'nonConstAnonymousNamespaceInt' is non-const and globally accessible, consider making it const [cppcoreguidelines-avoid-non-const-global-variables]
} // namespace
```

(There are also two check-nots where the line is off by one)

**Review by 5chmidti — APPROVED** [review]
While I'm leaning more towards 
> NOLINT in the code can grab the attention of a future developer and prompt a refactoring to avoid the pattern.

, it's an off-by-default option that provides a choice.

LGTM

**Review by HerrCai0907 — APPROVED** [review]

**Pull request merged**